### PR TITLE
Satellite map

### PR DESCRIPTION
**Pull request by pangal**
This branch adds functionality for creating a "Satellite map" of the generated world. This is currently an option, default false, toggled on with the --sat command when running.

Additional work may need to be done in selecting more appropriate biome colors and tuning the algorithm to display elevation / shading, to get a more realistic result.

Some sample outputs are attached here:

seed_35066_satellite
seed_53057_satellite

### CODE REVIEW

**Comment by esampson** [comment]
Looks pretty nice. It probably needs to be able to identify ocean that is cold enough to freeze and color that as well but beyond that it is probably just a matter of tweaking values.

File: worldengine/cli/main.py [hunk 1]
```diff
@@ -356,13 +363,18 @@ def main():
...
     export_options = parser.add_argument_group(
         "Export Options", "You can specify the formats you wish the generated output to be in. ")
-    export_options.add_argument("--export-format", dest="export_format", type=str,
-                                help="Export to a specific format such as BMP or PNG. " +
-                                "See http://www.gdal.org/formats_list.html for possible formats.",
-                                default="PNG")
-    export_options.add_argument("--export-datatype", dest="export_datatype", type=str,
-                                help="Type of stored data, e.g. uint16, int32, float32 etc.",
-                                default="uint16")
+    export_options.add_argument("--export-type", dest="export_type",
```
Comment (ftomassetti) [inline worldengine/cli/main.py]: Are these changes intended?

File: worldengine/cli/main.py [hunk 2]
```diff
@@ -381,7 +393,7 @@ def main():
     sys.setrecursionlimit(args.recursion_limit)
 
     if args.number_of_plates < 1 or args.number_of_plates > 100:
-        usage(error="Number of plates should be in [1, 100]")
+        usage(error="Number of plates should be a in [1, 100]")
```
Comment (ftomassetti) [inline worldengine/cli/main.py]: mmm perhaps you need to rebase the branch on master. I can do that for you if you want

**Comment by ftomassetti** [comment]
Thank you so much for sending this patch: it looks great! Now shadows are very beatiful and contribute a lot to make the map look real.

I just think there were a few changes in the master branch since you created yours so now I see some changes which I think are unrelated to you patch. Probably you need to rebase your branch on master. If you are not very familiar with git I would glad to do that for you.

Thanks a lot!

File: worldengine/cli/main.py (hunk 1 above)
Comment (pangal) [inline worldengine/cli/main.py]: No. I have no idea how they got in, git says I'm up to date with master but I'm thinking maybe I got some weirdly outdated version?

File: worldengine/cli/main.py (hunk 2 above)
Comment (pangal) [inline worldengine/cli/main.py]: I'm not sure what that means... my local repo says I'm up to date with master. I was puzzled by these changes too.

**Comment by pangal** [comment]
This is my first pull request. I'd be delighted to have you rebase with master; I'd also love it if you can point me to how I should be doing this! I tried rebasing it myself but it still says I'm up to date.

**Comment by ftomassetti** [comment]
Yeah, I tried to rebase it but it seems to be already up to date. Looking at the commits is like the first one ("Add 'satellite map' - a way to visualize the world as it may...") is the one containing all these differences.
Git is very mysterious from time to time...

Normally I would expect a rebase to fix this, but in this case I would fix it manually. Let me see what I can do.

**Comment by pangal** [comment]
I think I know the issue then. I was originally working on a cloned branch direct from your repo, but I had to manually copy my changed files after I officially forked this repo -- those changes from master must have happened then since I didn't try to pull from master before I copied. 

I can probably sort it out myself and make another pull request if you can't manually fix it easily.

**Comment by ftomassetti** [comment]
Ok, I updated the first commit and it is now in the branch pangal-satellite_map on this repository. If you want you can take it in your branch and push it on your fork. This would automatically update the pull-request. If you need help with that let me know.

My role as one of the maintainer is helping taking the more bits out of the way so that contributors like you can focus on providing great features like this one :)

**Comment by pangal** [comment]
Just tried to do this - it's still showing some things I didn't change in the diff.

**Comment by ftomassetti** [comment]
perhaps you have merged my branch while you should reset your branch to mine. This is because I "rewrote the history", so I changed the commits that were already part of your branch.

To do so you could for example to this:

```
# now the name "upstream" refer to this repository
git remote add upstream https://github.com/Mindwerks/worldengine.git
# copy the data from this repository on your computer
git fetch upstream
# move to the branch
git checkout satellite_map
# throw away all your stuff in the branch and make it exactly the same as the branch with 
# the same name on this repository. Note that this change only data on your computer not on your
# forked repository
git reset --hard upstream/satellite_map
# now sends the data to your forked repository: the pull request is updated automatically
git push origin satellite_map -f
```

File: worldengine/cli/main.py (hunk 1 above)
Comment (tcld) [inline worldengine/cli/main.py]: These are undone changes from a PR of mine that was pulled yesterday, I believe. Did you maybe do a rebase and merge in the wrong set of changes, i.e. the old code?

File: worldengine/cli/main.py (hunk 2 above)
Comment (tcld) [inline worldengine/cli/main.py]: Did you fork WorldEngine? If so, being up-to-date with "master" might mean that you are up-to-date with *your* fork, not the actual master.
(This again is something I recently changed.)

Comment (ftomassetti) [inline worldengine/cli/main.py]: No, I think the changes are part of the commit. I tried to checkout the branch from his repository and rebase it on origin/master but that did not work. I had to edit history, rebasing interactively and picking single changes from the first commit.

**Comment by tcld** [comment]
My first contribution was only a couple of weeks ago. I followed the guides on git to get started:
https://help.github.com/articles/fork-a-repo/
And it seems that worked out ok.

I guess I am still going to look over your code, even though things might be a bit messy right now?

File: worldengine/cli/main.py [hunk 3]
```diff
@@ -105,6 +107,8 @@ def generate_plates(seed, world_name, output_dir, width, height,
 
     # Generate images
     filename = '%s/plates_%s.png' % (output_dir, world_name)
+    # TODO calculate appropriate sea_level
+    sea_level = 1.0
```
Comment (tcld) [inline worldengine/cli/main.py]: This is more stuff that has been changed recently. I hope we find a way to clearly separate the actual new code from regressions or a lot of things might break.

File: worldengine/draw.py [hunk 4]
```diff
@@ -146,6 +198,9 @@ def set_pixel(self, x, y, color):
             color = (color[0], color[1], color[2], 255)
         self.pixels[x, y] = color
 
+    def get_pixel(self, x, y):
+        return self.pixels[x, y]
```
Comment (tcld) [inline worldengine/draw.py]: You can technically achieve this by calling ```get_item()```, i.e. "pixel-thingy[x, y]" - I don't know if that is the nicest habit, but it certainly works. So maybe this method can be left out.

File: worldengine/draw.py [hunk 5]
```diff
@@ -201,34 +268,424 @@ def draw_riversmap(world, target):
...
+
+
+def draw_satellite(world, target):
+    # Not sure what this first part is doing - copied from draw_greyscale_heightmap 
+    # I believe it is normalizing the heightmap values between 1 and 255?
+    # This function will use those values later, so this loop is kept in
+    min_elev_sea = None
+    max_elev_sea = None
+    min_elev_land = None
+    max_elev_land = None
```
Comment (tcld) [inline worldengine/draw.py]: You should copy this code from the most recent WorldEngine-code instead since there is a much faster way of doing this now.

File: worldengine/draw.py [hunk 6]
```diff
@@ -201,34 +268,424 @@ def draw_riversmap(world, target):
...
```
Comment (tcld) [inline worldengine/draw.py]: Again, recopy the code. ;) (It is just looking for min- and max-values in the elevation-array.)

File: worldengine/generation.py [hunk 7]
```diff
@@ -200,7 +200,7 @@ def generate_world(w, step):
 
     # Prepare sufficient seeds for the different steps of the generation
     rng = numpy.random.RandomState(w.seed)  # create a fresh RNG in case the global RNG is compromised (i.e. has been queried an indefinite amount of times before generate_world() was called)
-    sub_seeds = rng.randint(0, 4294967295, size=100)  # sys.maxsize didn't quite work
+    sub_seeds = rng.randint(0, 2147483647, size=100)  # sys.maxsize didn't quite work
```
Comment (tcld) [inline worldengine/generation.py]: This looks deliberate. numpy can digest a uint32 - what was your reasoning here?

**Comment by tcld** [comment]
I only breezed through the code, I don't have the time for a full check right now. And the regressions make it a bit hard to read anyway.

But: I love your results *and* your code.^^ Lots of comments, the code just looks tidy - this is a great point to start work for whoever might decide to do so.
(Of course that doesn't mean that I wouldn't want to see a couple small changes. Some I already mentioned. ;) )

File: worldengine/generation.py (hunk 7 above)
Comment (pangal) [inline worldengine/generation.py]: It wouldn't run on my 32-bit python install otherwise.

**Comment by pangal** [comment]
Thanks for the feedback all!

I started from scratch and created an entirely new pull request - https://github.com/Mindwerks/worldengine/pull/181 

I think it would be easier if we forget this one ever existed :)
